### PR TITLE
Graph

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/graph.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        '''Pass in a graph in the form of a dictionary with the nodes as
+        key, and a list of edges as value, for example
+        {1:[2, 3]},
+         2:[]},
+         3:[1]}
+        '''
+        if initial_graph is None:
+            self.graph = {}
+        else:
+            self.graph = initial_graph
```
Comment (nhuntwalker) [inline src/graph.py:16]: What if I pass in an object that isn't a dictionary? Like a float? Int? String?

File: src/graph.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    def __iter__(self):
+        return iter(self.graph)
+
+    def nodes(self):
+        '''return a list of all nodes in the graph'''
+        return self.graph.keys()
```
Comment (nhuntwalker) [inline src/graph.py:23]: Technically doesn't return a list in Python 3

File: src/graph.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    def edges(self):
+        '''returns a list of tuples, that are edges'''
+        return self.graph.values()
```
Comment (nhuntwalker) [inline src/graph.py:27]: This returns a list of neighbors for each node, not actual edges. An edge is a relationship between each node and its neighbors individually, e.g.:
for graph

```
g = {
    1: [2, 3, 4, 5],
    2: [4, 5],
    3: [],
    4: [6]
}
```

the printed edges should be **something** like

```
"1-2", "1-3", "1-4", "1-5", "2-4", "2-5", "4-6"
```

File: src/graph.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    def add_node(self, n):
+        '''takes node and adds it to the graph'''
+        if n in self.graph:
+            raise ValueError('Node already exists please try again')
+        else:
+            self.graph[n] = []
```
Comment (nhuntwalker) [inline src/graph.py:34]: This works, but there's actually a better way than using an `if...else...` block. Dictionaries have a method called `setdefault()` that will set a value on a given key if that key is not in the dictionary, and if it is in the dictionary it'll simply leave it alone.

File: src/graph.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    def add_edge(self, n1, n2):
+        '''adds a new edge to the graph connecting ‘n1’ and ‘n2’, if either n1 or n2 are not already
+         present in the graph, they are added.'''
+        if n2 not in self.graph:
+            self.graph[n2] = []
```
Comment (nhuntwalker) [inline src/graph.py:40]: there's a method for this

File: src/graph.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        if n in self.graph.keys():
+            self.graph.pop(n)
+        else:
+            raise ValueError('That node does not exist')
+
+    def del_edge(self, n1, n2):
+        '''deletes the edge connecting ‘n1’ and ‘n2’ from the graph, 
+        raises an error if no such edge exists'''
+        if (n1 in self.graph) and (n2 in self.graph):
+            self.graph[n1].remove(n2)
```
Comment (nhuntwalker) [inline src/graph.py:56]: What happens when nodes `n1` and `n2` are in the graph but there's no actual edge between the two?

File: src/graph.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        else:
+            raise ValueError('That node does not exist')
+
+    def has_node(self, n):
```
Comment (nhuntwalker) [inline src/graph.py:60]: You have this method. Why don't you use it in the rest of your graph whenever you're checking for membership?

File: src/graph.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        '''True if node ‘n’ is contained in the graph, False if not.'''
+        return n in self.graph
+
+    def neighbors(self, n):
+        '''returns the list of all nodes connected to ‘n’ by edges, raises an error if n is not in graph'''
+        if n in self.graph:
+            return self.graph.get(n)
```
Comment (nhuntwalker) [inline src/graph.py]: Why use `.get()` here? It's a dictionary right? Just do `self.graph[n]`

Comment (nhuntwalker) [inline src/graph.py:67]: Why use `.get()` here? It's a dictionary right? Just do `self.graph[n]`

File: src/graph.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        else:
+            raise ValueError('That node does not exist')
+
+    def adjacent(self, n1, n2):
+        '''returns True if there is an edge connecting n1 and n2, False if not,
+    raises an error if either of the supplied nodes are not in graph'''
+        if (n1 in self.graph) and (n2 in self.graph):
+            if n2 in self.graph[n1]:
+                return True
```
Comment (nhuntwalker) [inline src/graph.py:76]: Why not just `return n2 in self.graph[n1]`?

File: src/graph.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            else:
+                return False
+        else:
+            raise ValueError('That node does not exist')
```
Comment (nhuntwalker) [inline src/graph.py:80]: Which node doesn't exist?

File: src/graph.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+    def del_edge(self, n1, n2):
+        '''deletes the edge connecting ‘n1’ and ‘n2’ from the graph, 
+        raises an error if no such edge exists'''
+        if (n1 in self.graph) and (n2 in self.graph):
+            self.graph[n1].remove(n2)
+        else:
+            raise ValueError('That node does not exist')
```
Comment (nhuntwalker) [inline src/graph.py:58]: Which node doesn't exist?

File: src/graph.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+         2:[]},
+         3:[1]}
+        '''
+        if initial_graph is None:
+            self.graph = {}
+        else:
+            self.graph = initial_graph
+
+    def __iter__(self):
+        return iter(self.graph)
```
Comment (nhuntwalker) [inline src/graph.py:19]: Why do you need this? If your graph is a dictionary, it's already iterable.

File: src/test_graph.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    [[2, 3], [], [1]]
+)]
+
+
+@pytest.fixture
+def initial_graph():
+    graph = {1: [2, 3],
+             2: [],
+             3: [1]}
+    return graph
```
Comment (nhuntwalker) [inline src/test_graph.py:16]: You only really ever test one type of graph, or an empty one. You should potentially come up with alternative ones to test.

File: src/graph.py (hunk 1 above)
Comment (crashtack) [reply src/graph.py:16]: then the crap out for the crap they passed in

Comment (ghost) [reply src/graph.py:16]: Done

File: src/graph.py (hunk 8 above)
Comment (ghost) [reply src/graph.py:67]: Done

File: src/graph.py (hunk 9 above)
Comment (ghost) [reply src/graph.py:76]: Done

File: src/graph.py (hunk 11 above)
Comment (ghost) [reply src/graph.py:58]: Half done

File: src/graph.py (hunk 12 above)
Comment (ghost) [reply src/graph.py:19]: TypeError: argument of type 'Graph' is not iterable

**Pull request merged**